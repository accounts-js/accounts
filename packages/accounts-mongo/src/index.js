// @flow

import { get } from 'lodash';
import { encryption } from '@accounts/server';
import type {
  CreateUserType,
  UserObjectType,
  SessionType,
} from '@accounts/common';
import { ObjectID } from 'mongodb';

export type MongoOptionsType = {
  collectionName: string,
  sessionCollectionName: string,
  timestamps: {
    createdAt: string,
    updatedAt: string,
  }
};

export type MongoUserObjectType = {
  username?: string,
  profile?: Object,
  services: {
    password?: {
      bcrypt: string,
    },
  },
  emails?: [{
    address: string,
    verified: boolean,
  }],
};

const toMongoID = (objectId) => {
  if (typeof objectId === 'string') {
    return new ObjectID(objectId);
  }

  return objectId;
};

class Mongo {
  options: MongoOptionsType;
  // TODO definition for mongodb connection object
  db: any;
  collection: any;
  sessionCollection: any;

  constructor(db: any, options: MongoOptionsType) {
    const defaultOptions = {
      collectionName: 'users',
      sessionCollectionName: 'sessions',
      timestamps: {
        createdAt: 'createdAt',
        updatedAt: 'updatedAt',
      },
    };
    this.options = { ...defaultOptions, ...options };
    if (get(db, 'constructor.define.name') !== 'Db') {
      throw new Error('A valid database connection object is required');
    }
    this.db = db;
    this.collection = this.db.collection(this.options.collectionName);
    this.sessionCollection = this.db.collection(this.options.sessionCollectionName);
  }

  async setupIndexes(): Promise<void> {
    await this.collection.createIndex('username', { unique: 1, sparse: 1 });
    await this.collection.createIndex('emails.address', { unique: 1, sparse: 1 });
  }

  async createUser(options: CreateUserType): Promise<string> {
    const user: MongoUserObjectType = {
      services: {},
      profile: {},
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    };
    if (options.password) {
      user.services.password = { bcrypt: await encryption.hashPassword(options.password) };
    }
    if (options.username) {
      user.username = options.username;
    }
    if (options.email) {
      user.emails = [{ address: options.email.toLowerCase(), verified: false }];
    }
    if (options.profile) {
      user.profile = options.profile;
    }
    const ret = await this.collection.insertOne(user);
    return ret.ops[0]._id;
  }

  async findUserById(userId: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ _id: toMongoID(userId) });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByEmail(email: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ 'emails.address': email.toLowerCase() });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findUserByUsername(username: string): Promise<?UserObjectType> {
    const user = await this.collection.findOne({ username });
    if (user) {
      user.id = user._id;
    }
    return user;
  }

  async findPasswordHash(userId: string): Promise<?string> {
    const user = await this.findUserById(toMongoID(userId));
    if (user) {
      return user.services.password.bcrypt;
    }
    return null;
  }

  async addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $addToSet: {
        emails: {
          address: newEmail.toLowerCase(),
          verified,
        },
      },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async removeEmail(userId: string, email: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $pull: { emails: { address: email.toLowerCase() } },
      $set: { [this.options.timestamps.updatedAt]: Date.now() },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setUsername(userId: string, newUsername: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        username: newUsername,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setPasssword(userId: string, newPassword: string): Promise<void> {
    const ret = await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        'services.password.bcrypt': await encryption.hashPassword(newPassword),
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    if (ret.result.nModified === 0) {
      throw new Error('User not found');
    }
  }

  async setProfile(userId: string, profile: Object): Promise<Object> {
    await this.collection.update({ _id: toMongoID(userId) }, {
      $set: {
        profile,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
    const user = await this.findUserById(userId);
    return user && user.profile;
  }

  async createSession(userId: string, ip: string, userAgent: string): Promise<string> {
    const ret = await this.sessionCollection.insertOne({
      userId,
      userAgent,
      ip,
      valid: true,
      [this.options.timestamps.createdAt]: Date.now(),
      [this.options.timestamps.updatedAt]: Date.now(),
    });
    return ret.ops[0]._id;
  }

  async updateSession(sessionId: string, ip: string, userAgent: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        ip,
        userAgent,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  async invalidateSession(sessionId: string): Promise<void> {
    await this.sessionCollection.update({ _id: toMongoID(sessionId) }, {
      $set: {
        valid: false,
        [this.options.timestamps.updatedAt]: Date.now(),
      },
    });
  }

  findSessionById(sessionId: string): Promise<?SessionType> {
    return this.sessionCollection.findOne({ _id: toMongoID(sessionId) });
  }
}

export default Mongo;
