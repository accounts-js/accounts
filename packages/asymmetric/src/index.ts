import AccountsAsymmetric from './accounts-asymmetric';

export default AccountsAsymmetric;
export { AccountsAsymmetric };
