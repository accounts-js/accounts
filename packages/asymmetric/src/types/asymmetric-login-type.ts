export interface AsymmetricLoginType {
  publicKey: string;
  signature: string;
  signatureAlgorithm: 'sha256' | 'sha512';
  signatureFormat: 'hex' | 'base64';
  payload: string;
}
