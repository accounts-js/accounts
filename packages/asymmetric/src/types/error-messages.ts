export interface ErrorMessages {
  userNotFound: string;
  verificationFailed: string;
}
