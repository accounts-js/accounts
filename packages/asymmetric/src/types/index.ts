export * from './asymmetric-login-type';
export * from './public-key-type';
