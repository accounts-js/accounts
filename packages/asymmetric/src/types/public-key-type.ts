export interface PublicKeyType {
  key: string;
  encoding: 'utf8' | 'base64' | 'hex';
  format: 'pem' | 'der';
  type: 'pkcs1' | 'spki';
}
