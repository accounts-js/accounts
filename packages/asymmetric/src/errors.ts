import { ErrorMessages } from './types';

export const errors: ErrorMessages = {
  userNotFound: 'User not found',
  verificationFailed: 'Failed to verify signature',
};
