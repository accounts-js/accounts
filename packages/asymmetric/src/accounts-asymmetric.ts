import * as crypto from 'crypto';

import { AuthenticationService, DatabaseInterface } from '@accounts/types';
import { AccountsServer } from '@accounts/server';
import forge from 'node-forge';

import { AsymmetricLoginType, PublicKeyType, ErrorMessages } from './types';
import { errors } from './errors';

export interface AccountsAsymmetricOptions {
  errors?: ErrorMessages;
}

const defaultOptions = {
  errors,
};

export default class AccountsAsymmetric implements AuthenticationService {
  public serviceName = 'asymmetric';
  public server!: AccountsServer;
  private db!: DatabaseInterface;
  private options: AccountsAsymmetricOptions & typeof defaultOptions;

  constructor(options: AccountsAsymmetricOptions = {}) {
    this.options = { ...defaultOptions, ...options };
  }

  public setStore(store: DatabaseInterface) {
    this.db = store;
  }

  public async updatePublicKey(
    userId: string,
    { key, ...params }: PublicKeyType
  ): Promise<boolean> {
    try {
      await this.db.setService(userId, this.serviceName, { id: key, key, ...params });
      return true;
    } catch (e) {
      return false;
    }
  }

  public async authenticate(params: AsymmetricLoginType): Promise<any> {
    const user = await this.db.findUserByServiceId(this.serviceName, params.publicKey);

    if (!user) {
      throw new Error(this.options.errors.userNotFound);
    }

    try {
      const verify = crypto.createVerify(params.signatureAlgorithm);
      verify.write(params.payload);
      verify.end();

      const publicKeyParams: PublicKeyType = (user.services as any)[this.serviceName];

      const pemText =
        publicKeyParams.format === 'pem' ? publicKeyParams.key : this.derToPem(publicKeyParams);

      const isVerified = verify.verify(pemText, params.signature, params.signatureFormat);

      return isVerified ? user : null;
    } catch (e) {
      throw new Error(this.options.errors.verificationFailed);
    }
  }

  private derToPem(publicKeyParams: PublicKeyType): string {
    let derKey;

    switch (publicKeyParams.encoding) {
      case 'utf8': {
        derKey = forge.util.decodeUtf8(publicKeyParams.key);
        break;
      }
      case 'hex': {
        derKey = forge.util.hexToBytes(publicKeyParams.key);
        break;
      }
      case 'base64':
      default: {
        derKey = forge.util.decode64(publicKeyParams.key);
        break;
      }
    }

    const asnObj = forge.asn1.fromDer(derKey);
    const publicKey = forge.pki.publicKeyFromAsn1(asnObj);
    return forge.pki.publicKeyToPem(publicKey);
  }
}
