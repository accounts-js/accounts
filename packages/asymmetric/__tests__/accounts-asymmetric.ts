import * as crypto from 'crypto';

import { PublicKeyType } from '../src/types';
import { AccountsAsymmetric } from '../src';

describe('AccountsAsymmetric', () => {
  it('should update the public key', async () => {
    const { publicKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 512,
      publicKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
      privateKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
    });

    const service = new AccountsAsymmetric();

    const setService = jest.fn(() => Promise.resolve(null));
    service.setStore({ setService } as any);

    const publicKeyParams: PublicKeyType = {
      key: publicKey,
      encoding: 'base64',
      format: 'pem',
      type: 'spki',
    };

    const res = await service.updatePublicKey('123', publicKeyParams);

    expect(res).toBeTruthy();
    expect(setService).toHaveBeenCalledWith('123', service.serviceName, {
      id: publicKey,
      ...publicKeyParams,
    });
  });

  it('throws error when the user is not found by public key', async () => {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 512,
      publicKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
      privateKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
    });
    const payload = 'some data to sign';

    const sign = crypto.createSign('SHA256');
    sign.write(payload);
    sign.end();
    const signature = sign.sign(privateKey, 'hex');

    const service = new AccountsAsymmetric();

    const findUserByServiceId = jest.fn(() => Promise.resolve(null));
    service.setStore({ findUserByServiceId } as any);

    await expect(
      service.authenticate({
        signature,
        payload,
        publicKey,
        signatureAlgorithm: 'sha512',
        signatureFormat: 'hex',
      })
    ).rejects.toMatchSnapshot();
  });

  it('should return null when signature is invalid', async () => {
    const { publicKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 512,
      publicKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
      privateKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
    });
    const payload = 'some data to sign';

    const service = new AccountsAsymmetric();

    const publicKeyParams: PublicKeyType = {
      key: publicKey,
      encoding: 'base64',
      format: 'pem',
      type: 'pkcs1',
    };

    const user = {
      id: '123',
      services: {
        [service.serviceName]: publicKeyParams,
      },
    };
    const findUserByServiceId = jest.fn(() => Promise.resolve(user));
    service.setStore({ findUserByServiceId } as any);

    const userFromService = await service.authenticate({
      signature: 'some signature',
      payload,
      publicKey: publicKey,
      signatureAlgorithm: 'sha256',
      signatureFormat: 'hex',
    });

    expect(userFromService).toBeNull();
  });

  it('should return user when verification is successful', async () => {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 512,
      publicKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
      privateKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
    });
    const payload = 'some data to sign';

    const sign = crypto.createSign('sha256');
    sign.write(payload);
    sign.end();
    const signature = sign.sign(privateKey, 'hex');

    const service = new AccountsAsymmetric();

    const publicKeyParams: PublicKeyType = {
      key: publicKey,
      encoding: 'base64',
      format: 'pem',
      type: 'pkcs1',
    };

    const user = {
      id: '123',
      services: {
        [service.serviceName]: publicKeyParams,
      },
    };
    const findUserByServiceId = jest.fn(() => Promise.resolve(user));
    service.setStore({ findUserByServiceId } as any);

    const userFromService = await service.authenticate({
      signature,
      payload,
      publicKey: publicKey,
      signatureAlgorithm: 'sha256',
      signatureFormat: 'hex',
    });

    expect(userFromService).toEqual(user);
  });

  it('should return user when verification is successful using der format', async () => {
    const { publicKey: publicKeyBuffer, privateKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 512,
      publicKeyEncoding: {
        type: 'pkcs1',
        format: 'der',
      },
      privateKeyEncoding: {
        type: 'pkcs1',
        format: 'pem',
      },
    });
    const publicKey = publicKeyBuffer.toString('base64');
    const payload = 'some data to sign';

    const sign = crypto.createSign('sha256');
    sign.write(payload);
    sign.end();
    const signature = sign.sign(privateKey, 'hex');

    const service = new AccountsAsymmetric();

    const publicKeyParams: PublicKeyType = {
      key: publicKey,
      encoding: 'base64',
      format: 'der',
      type: 'pkcs1',
    };

    const user = {
      id: '123',
      services: {
        [service.serviceName]: publicKeyParams,
      },
    };
    const findUserByServiceId = jest.fn(() => Promise.resolve(user));
    service.setStore({ findUserByServiceId } as any);

    const userFromService = await service.authenticate({
      signature,
      payload,
      publicKey,
      signatureAlgorithm: 'sha256',
      signatureFormat: 'hex',
    });

    expect(userFromService).toEqual(user);
  });
});
