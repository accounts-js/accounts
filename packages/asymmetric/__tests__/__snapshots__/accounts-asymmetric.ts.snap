// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsAsymmetric throws error when the user is not found by public key 1`] = `[Error: User not found]`;
